service/timestreamwrite/api_client.go

```diff
@@ -241,6 +261,84 @@ func addRetryMiddlewares(stack *middleware.Stack, o Options) error {
 	return retry.AddRetryMiddlewares(stack, mo)
 }
 
+// resolves EnableEndpointDiscovery configuration
+func resolveEnableEndpointDiscoveryFromConfigSources(cfg aws.Config, o *Options) error {
+	if len(cfg.ConfigSources) == 0 {
+		return nil
+	}
+	value, found, err := internalConfig.ResolveEnableEndpointDiscovery(context.Background(), cfg.ConfigSources)
+	if err != nil {
+		return err
+	}
+	if found {
+		o.EnableEndpointDiscovery = value
+	}
+	return nil
+}
+
+// resolves endpoint cache on client
+func resolveEndpointCache(c *Client) {
+	c.endpointCache = internalEndpointDiscovery.NewEndpointCache(internalEndpointDiscovery.DefaultEndpointCacheSize)
+}
+
+func resolveEnableEndpointDiscovery(o *Options) {
+	if len(o.EnableEndpointDiscovery) != 0 {
+		return
+	}
+	o.EnableEndpointDiscovery = aws.EndpointDiscoveryAuto
+}
+
+func (c *Client) handleEndpointDiscoveryFromService(ctx context.Context, input *DescribeEndpointsInput, key string, opt internalEndpointDiscovery.DiscoverEndpointOptions) (internalEndpointDiscovery.Endpoint, error) {
+	output, err := c.DescribeEndpoints(ctx, input, func(o *Options) {
+		o.EndpointOptions.DisableHTTPS = opt.DisableHTTPS
+		o.Logger = opt.Logger
+		if len(opt.EndpointUsedForDiscovery) != 0 {
+			o.EndpointResolver = EndpointResolverFromURL(opt.EndpointUsedForDiscovery)
+		}
+
+	})
+	if err != nil {
+		return internalEndpointDiscovery.Endpoint{}, err
+	}
+
+	endpoint := internalEndpointDiscovery.Endpoint{}
+	endpoint.Key = key
+
+	for _, e := range output.Endpoints {
+		if e.Address == nil {
+			continue
+		}
+		address := *e.Address
+
+		var scheme string
+		if idx := strings.Index(address, "://"); idx != -1 {
+			scheme = address[:idx]
+		}
+		if len(scheme) == 0 {
+			scheme = "https"
+			if opt.DisableHTTPS {
+				scheme = "http"
+			}
+			address = fmt.Sprintf("%s://%s", scheme, address)
+		}
+
+		cachedInMinutes := e.CachePeriodInMinutes
+		u, err := url.Parse(address)
+		if err != nil {
+			continue
+		}
+
+		addr := internalEndpointDiscovery.WeightedAddress{
+			URL:     u,
+			Expired: time.Now().Add(time.Duration(cachedInMinutes) * time.Minute),
```

Comment:
This probably a spot where we want to strip monotonic clock readings off to make this safe if an application is on a system that is hibernated and then resumed. By calling `t.Round(0)`